ARCs/arc-0060.md

```diff
@@ -0,0 +1,330 @@
+
+##### Simple Message Data JSON Schema
+
+The JSON schema for simple message signing. The `StdData` **MUST** be an object of this schema.
+
+> The signing data JSON schema is inspired by the schema proposed with <a href="https://eips.ethereum.org/EIPS/eip-712">EIP-712: Typed structured data hashing and signing proposal</a>.
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "ARC60Domain": {
+      "type": "string",
+      "description": "The ARC-60 domain separator"
+    },
+    "bytes": {
+      "type": "string",
+      "description": "Byte string being signed. Used with ScopeType==MSGSIG."
+    },
+  "required": ["ARC60Domain", "bytes"]
+}
+```
+
+- The `ARC60Domain` object indicates the domain separator to be used for signing. It **SHOULD** be set to `"arc60"` for simple bytes. Algorand transaction domain separators `TX` and `TG` **MUST** be forbidden.
+- The `bytes` object is the byte array of data being signed. It is a simple string in case of `ScopeType==MSGSIG`, a more complex schema **SHOULD** be used otherwise. It cannot be a valid Algorand transaction. The value of `bytes` **MUST NOT** be prepended with a known domain separator `TX` or `TG`.
+
+> Algorand domain separators can be found in the <a href="https://github.com/algorandfoundation/specs/blob/master/dev/crypto.md#domain-separation">Algorand specs</a> and in the <a href="https://github.com/algorand/go-algorand/blob/master/protocol/hash.go#L21">Go type HashID</a>.
+
+For example, a valid object that takes as bytes an extra property field is:
+
+```json
+{
+  "ARC60Domain" : "arc60",
+  "bytes" : "ARC-60 is awesome"
+}
+```
+
+#### Error interface `SignDataError`
+
+The `SignDataError` object extends the `SignTxnsError` defined in [ARC-1](./arc-0001.md).
+
+```ts
+export interface SignDataError extends Error {
+  code: number;
+  data?: any;
+  failingSignData: (StdData | null);
+}
+```
+
+`SignDataError` uses the same error codes of `SignTxnsError` as well as the following codes:
+
+| Status Code | Name | Description |
+| --- | --- | --- |
+| 4600 | Invalid scope | The scope of the signing action is not recognized by the wallet |
+| 4601 | Invalid schema | The schema does not comply with ARC-60 requirements |
+| 4602 | Invalid encoding | The canonical JSON cannot be decoded with the given encoding |
+
+### Semantic Requirements
+
+The call `signData(data, metadata, signer)` **MUST** either return the signed data `ret` or reject the call throwing an error `err`.
+
+> Following [ARC-1](./arc-0001.md) terminology, in this ARC the term **Rejecting** means throwing an error with `err.code=4300`.
+
+Upon calling `signData(data, metadata, signer)`:
+
+- the `data`, `metadata`, and `signer` **MUST NOT** be `null`, otherwise the wallet **MUST** reject the call.
+- `data` **MUST** be validated with respect to the JSON schema `metadata.schema`. If the validation fails, the call **MUST** be rejected with a `4300` error.
```

Comment:
According to the table above an invalid schema is a 4601 error code, or is this a different schema?

Reply:
Error code 4601 means that the JSON schema does not comply with ARC-60 (the required fields `ARC60Domain` and `bytes` are missing).

In this case, we are failing because `data` is not a valid object according to the schema `metadata.schema`.